tensorflow/compiler/xla/service/gpu/done_event_insertion.cc

```diff
@@ -0,0 +1,49 @@
+#include "tensorflow/compiler/xla/service/gpu/done_event_insertion.h"
+
+#include "tensorflow/compiler/xla/service/hlo_casting_utils.h"
+#include "tensorflow/compiler/xla/service/hlo_instruction.h"
+#include "tensorflow/compiler/xla/service/hlo_instructions.h"
+#include "tensorflow/compiler/xla/service/hlo_opcode.h"
+
+namespace xla {
+namespace {
+const char* const kDoneEvent = "__builtin$DoneEvent";
+// An element makes the liveness analysis lift the kernel as early as possible.
+const Shape DummyShape = ShapeUtil::MakeScalarShape(S32);
+
+Status AddDoneEvent(HloInstruction* inst, size_t index) {
+  if (inst->opcode() == HloOpcode::kBitcast) {
+    for (HloInstruction* op : inst->mutable_operands()) {
+      TF_RETURN_IF_ERROR(AddDoneEvent(op, index));
+    }
+  } else if (inst->opcode() == HloOpcode::kGetTupleElement) {
+    HloInstruction* tuple = inst->mutable_operand(0);
+    if (tuple->opcode() == HloOpcode::kTuple) {
+      TF_RETURN_IF_ERROR(
+          AddDoneEvent(tuple->mutable_operand(inst->tuple_index()), index));
+    } else {
+      // Instructions like fusion generates a tuple
+      TF_RETURN_IF_ERROR(AddDoneEvent(inst->mutable_operand(0), index));
+    }
+  } else {
+    auto done_event = Cast<HloCustomCallInstruction>(
+        inst->parent()->AddInstruction(HloInstruction::CreateCustomCall(
+            DummyShape, {inst}, kDoneEvent, std::to_string(index))));
+    done_event->set_custom_call_has_side_effect(true);
+  }
+  return Status::OK();
+}
+};  // namespace
+
+StatusOr<bool> HloDoneInsertion::Run(HloModule* module) {
+  bool changed = false;
+
+  HloComputation* entry = module->entry_computation();
+  HloInstruction* tuple = entry->root_instruction();
```

Comment:
I am not very familiar with HLO. After searching on the internet, I find that `root_instruction` is used for HloModule's output. Is this the reason that you only consider this instruction in the whole  HloModule?

Reply:
yes as the name indicates it is the entry. All outputs are from this function so we search from it until an operator whose input and output do not alias.

Reply:
OK, thx.